File: src/main/java/ar/edu/utn/frba/dds/domain/comunidades/Comunidad.java
package ar.edu.utn.frba.dds.domain.comunidades;

import ar.edu.utn.frba.dds.domain.MediosDeComunicacion.Notificacion;
import ar.edu.utn.frba.dds.domain.incidentes.Incidente;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import ar.edu.utn.frba.dds.domain.incidentes.TipoFiltrado;

import javax.persistence.Column;
import javax.persistence.Entity;
import javax.persistence.GeneratedValue;
import javax.persistence.GenerationType;
import javax.persistence.Id;
import javax.persistence.JoinColumn;
import javax.persistence.JoinTable;
import javax.persistence.ManyToMany;
import javax.persistence.ManyToOne;
import javax.persistence.OneToMany;
import javax.persistence.Table;
import javax.persistence.Transient;
import lombok.Getter;
import lombok.Setter;

@Entity
@Table(name = "Comunidades")
@Getter @Setter
public class Comunidad{
  private List<Miembro> administradores;
  @ManyToMany
  @JoinTable(name = "miembros_por_comunidad",
      joinColumns = @JoinColumn(name = "miembro_id", referencedColumnName = "id"),
      inverseJoinColumns = @JoinColumn(name = "comunidad_id", referencedColumnName = "id")
  )
  private List<Miembro> miembros;
  @OneToMany
  private List<Incidente> incidentes;
  @Column(name = "nombre")
  private String nombre;

  public Comunidad() {
    this.nombre = "HOLA";
    this.administradores = new ArrayList<>();
    this.miembros = new ArrayList<>();
    this.incidentes = new ArrayList<>();
  }

   //esto s para fijarse el estado de los incidentes abiertos
  public void imprimirIncidentes(TipoFiltrado tipoFiltrado) {
    for (Incidente incidente : incidentes) {
      if (tipoFiltrado == TipoFiltrado.SOLO_ABIERTOS && incidente.getAbierto()) {
        System.out.println(incidente);
      } else if (tipoFiltrado == TipoFiltrado.SOLO_CERRADOS && !incidente.getAbierto()) {
        System.out.println(incidente);
      } else if (tipoFiltrado == TipoFiltrado.TODOS) {
        System.out.println(incidente);
      }
    }
  }
  
/*



  public void ingresarServicioNuevo(Establecimiento establecimiento, Miembro administrador, String nombre, String descripcion, int cantidad) throws NoEsAdministradorExcepcion {
    this.verificarQueEsAdministrador(administrador);
    ServicioComunitario servicio = new ServicioComunitario();
    servicio.setNombre(nombre);
    servicio.setDescripcion(descripcion);
    PrestacionDeServicio prestacionDeServicio = new PrestacionDeServicio(servicio,cantidad);
    prestacionDeServicio.setFunciona(false);
    establecimiento.agregarPrestaciones(prestacionDeServicio);
  }

  public void verificarQueEsAdministrador(Miembro administrador) throws NoEsAdministradorExcepcion{
    if (!this.administradores.contains(administrador)) {
      throw new NoEsAdministradorExcepcion();
    }
  }
*/
  public void agregarUsuarios(Miembro... miembros) {
    this.miembros.addAll(Arrays.asList(miembros));
  }

  public void agregarAdministradores(Miembro... administradores) {
    this.administradores.addAll(Arrays.asList(administradores));
  }

  public void removerUsuarios(Miembro... miembros) {
    for (Miembro value : miembros) {
      this.miembros.remove(value);
    }
  }
  public void removerAdministradores(Miembro... administradores) {
      for (Miembro value : administradores){
        this.administradores.remove(value);
      }
  }



  public void notificarMiembros(Incidente incidente){
    getIncidentes().add(incidente);
    Notificacion notificacion = new Notificacion(incidente);
    this.miembros.stream().filter(miembro -> miembro.getUsuario() != incidente.getAbridor().getUsuario())
                          .forEach(m->m.getMedioDeNotificacion().evaluarEnvioDeNotificacion(notificacion));
    System.out.println("Se ha enviado notificacion al WhatssApp - ");
  }

  public void cerrarIncidente(Miembro autor,Incidente incidente){
    incidente.meCierro(autor);
    this.notificarMiembros(incidente);
  }

}

File: src/main/ts/ar/edu/utn/frba/dds/domain/comunidades/Comunidad.ts
import {Column, Entity, PrimaryGeneratedColumn } from "typeorm";
import {Puntaje} from "./gradosDeConfianza/Puntaje";

@Entity()
export class Comunidad {
    @PrimaryGeneratedColumn()
    id: number

    administradores: Array<Miembro> = new Array<Miembro>()
    miembros: Array<Miembro> = new Array<Miembro>()

    propuestasFusion: Array<PropuestaFusion> = new Array<PropuestaFusion>()

    incidentes: Array<Incidente> = new Array<Incidente>()

    @Column()
    nombre: string

    @Column()
    descripcion: string

    puntaje: Puntaje

    agregarUsuarios (... miembros: Miembro) {
        this.miembros.push(miembros)
        for (let miembro in miembros) {
            miembro.comunidades.push(this)
        }
    }

    removerUsuarios (... miembros: Miembro) {
        for (let miembro in miembros) {
            const index = this.miembros.indexOf(miembro)
            this.miembros.splice(index, 1)
        }
    }

    cantidadDeMiembros(): number {
        return this.miembros.length
    }

    cantidadDeIncidentes(): number {
        return this.incidentes.length
    }

    notificarMiembros (incidente: Incidente) {
        this.incidentes.push(incidente)
        let notificacion: Notificacion = new Notificacion()
        notificacion.crearNotificacion(incidente)
        this.miembros.filter((miembro: Miembro) =>
            miembro.usuario != incidente.abridor.usuario
        ).forEach((miembro: Miembro) =>
            miembro.medioDeNotificacion.evaluarEnvioDeNotificacion(notificacion)
        )
    }

    cerrarIncidente (autor: Miembro, incidente: Incidente) {
        incidente.meCierro(autor)
        this.notificarMiembros(incidente)
    }

    comunidadPuntaje(): ComunidadPuntaje {
        return new ComunidadPuntajeBuilder.conId(this.id).conPuntaje(this.puntaje.valor).conMiembros(this.miembros).construir()
    }

    comunidadFusionable(): ComunidadFusionable {
        return new ComunidadFusionableBuilder().conId(this.id).conIncidentes(this.incidentes).conEstablecimientos(establecimientosDeIncidentes())
            .conServicios(serviciosDeIncidentes()).conUsuarios(this.miembros).conPropuestasAnteriores(this.propuestasFusion)
            .conGradoDeConfianza(this.puntaje.valor).construir()
    }

    establecimientosDeIncidentes(): Array<Establecimiento> {
        this.incidentes.map(Incidente::establecimiento)
    }

    serviciosDeIncidentes(): Array<PrestacionDeServicio> {
        this.incidentes.map(Incidente::prestacionDeServicio)
    }

    actualizarPuntajes (comunidadPuntaje: ComunidadPuntaje) {
        this.puntaje.valor(comunidadPuntaje.puntaje)
        actualizarPuntajesMiembros(comunidadPuntaje)
    }

    actualizarPuntajesMiembros (comunidadPuntaje: ComunidadPuntaje) {
        comunidadPuntaje.miembros.forEach((miembroPuntaje: MiembroPuntaje) =>
            this.miembros.forEach((miembro: Miembros) => {
                    if (miembro.id == miembroPuntaje.id) {
                        miembro.puntaje.valor(miembroPuntaje.puntaje)
                    }
                }
            )
        )
    }
}